pkg/api/rest/delete.go

```diff
@@ -69,14 +69,13 @@ func BeforeDelete(strategy RESTDeleteStrategy, ctx api.Context, obj runtime.Obje
 		// only a shorter grace period may be provided by a user
 		if options.GracePeriodSeconds != nil {
 			period := int64(*options.GracePeriodSeconds)
-			if period > *objectMeta.DeletionGracePeriodSeconds {
-				return false, true, nil
+			newDeletionTimestamp := unversioned.NewTime(unversioned.Now().Add(time.Second * time.Duration(*options.GracePeriodSeconds)))
```

Comment:
There are two cases for a user invoking two deletes in order (30s and 15s).

1. The owner of graceful (in this case, for a pod, the kubelet) sees the first request, acts on it, then sees the second
2. The owner of graceful sees both requests at the same time (or within the same interval)

For the first case, the graceful deleter should trigger a 30s delete, then when it sees a shorter interval, needs to make a decision in its own timeframe (in the causal sense) about whether to terminate or shorten.  In the second case, the initial deletion would be 15s.